client/shared/src/mem.rs

```diff
@@ -12,18 +12,56 @@ const fn MiB(n: usize) -> usize {
 
 const STACK_ALIGNMENT: usize = 1024 * 63;
 
+/// The size of the stack. Its start is at address 0
 pub const GRAPHIC_STACK_SIZE: usize = MiB(4) + STACK_ALIGNMENT;
+
+/// The address of the Allocator structures (size: 1MiB)
 pub const ALLOCATOR_AREA_START: usize = GRAPHIC_STACK_SIZE;
-pub const SHARED_BUFFER_AREA_START: usize = ALLOCATOR_AREA_START + MiB(1);
+
+/// The address memory communication area. (size: 1MiB)
+/// It contains data needed for communication between main thread and logic thread.
+pub const COMMUNICATION_MEMORY_START: usize = ALLOCATOR_AREA_START + MiB(1);
+
+/// The address of the double buffer (size: target dependent)
+pub const SHARED_BUFFER_AREA_START: usize =
+    COMMUNICATION_MEMORY_START + core::mem::size_of::<CommunicationMemory>();
+
+/// The logic heap address (size: 32MiB)
 pub const LOGIC_ALLOCATION_AREA_START: usize =
-    SHARED_BUFFER_AREA_START + std::mem::size_of::<Buffer>();
+    SHARED_BUFFER_AREA_START + core::mem::size_of::<Buffer>();
+
+/// The graphics heap address (size: 32MiB)
 pub const GRAPHICS_ALLOCATION_AREA_START: usize = LOGIC_ALLOCATION_AREA_START + MiB(32);
+
+/// The start of unbounded shared memory (size: unbounded)
 pub const SHARED_ALLOCATION_AREA_START: usize = GRAPHICS_ALLOCATION_AREA_START + MiB(32);
 
 pub fn get_double_buffer() -> &'static mut Buffer {
     unsafe { &mut *(SHARED_BUFFER_AREA_START as *mut Buffer) }
 }
 
+pub struct CommunicationMemory {
+    elapsed_ms: u64,
+    last_elapsed_ms: u64,
+}
+
+impl CommunicationMemory {
+    pub unsafe fn get() -> &'static Self {
+        &*(COMMUNICATION_MEMORY_START as *const Self)
+    }
+
+    pub unsafe fn get_mut() -> &'static mut Self {
+        &mut *(COMMUNICATION_MEMORY_START as *mut Self)
+    }
+
+    pub fn wait_for_main_thread_notify(&mut self) {
+        self.last_elapsed_ms = self.elapsed_ms;
+        while self.last_elapsed_ms == self.elapsed_ms {
```

Comment:
This seems fishy to me, could you explain, what that does?
this looks like it might wait too long

Reply:
The `last_elapsed_ms` is a field that is not touched by the main thread (probably I should add documentation). The `elapsed_ms`-field is synchronized by the main-thread. You would need an atomic read here, that's an issue I will fix soon. So this loop waits until wake up and then tests if the `elapsed_ms` field has changed.

Reply:
But is it the elapsed_ms the total elapsed time, or the time elapsed since the last sync?
That should at least be documented, or the variable could be renamed to express that it represents a continuous time